### PR TITLE
added function

### PR DESCRIPTION
**Pull request by BrandonMurnane**

### CODE REVIEW

File: refocus-bdk-client.js [hunk 1]
```diff
@@ -707,6 +707,35 @@ module.exports = (config) => {
         roomObject, TOKEN);
     }, // updateRoomName
 
+    getOrInitializeBotData: (room,botName,dataName,defaultValue) => {
```
Comment (tausifmuzaffar) [inline refocus-bdk-client.js]: have we started testing in BDK?

**Review by tausifmuzaffar — COMMENTED** [review]
have we started testing in BDK?

File: refocus-bdk-client.js [hunk 2]
```diff
@@ -707,6 +708,60 @@ module.exports = (config) => {
...
+              return botDataValue;
+            }
+            let newData = defaultValue;
+            if (newData && typeof newData !== 'string') {
+              newData = serialize(newData);
+            }
+            const botData = {
+              'name': dataName,
+              'roomId': parseInt(room, 10),
+              'botId': bot,
```
Comment (kriscfoster) [inline refocus-bdk-client.js]: undefined from test

**Comment by kriscfoster** [comment]
@BrandonMurnane conflicts

**Review by kriscfoster — APPROVED** [review]

**Comment by kriscfoster** [comment]
Just the function name in the readme and looks good

**Review by kriscfoster — CHANGES_REQUESTED** [review]
New function names in the readme

File: refocus-bdk-client.js [hunk 3]
```diff
@@ -720,6 +721,44 @@ module.exports = (config) => {
...
+            }
+            const botData = {
+              'name': dataName,
+              'roomId': parseInt(room, 10),
+              'botId': botName,
+              'value': newData
+            };
+            return genericPost(`${SERVER}${API}${BOTDATA_ROUTE}`,
+              botData, TOKEN)
+              .then(() => newData);
```
Comment (cassiodias) [inline refocus-bdk-client.js]: What happens if there is something wrong with this post? Should the client always rely on 'newData'?

File: tests/getOrInitializeBotData.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+const bdk = rewire('../refocus-bdk-client.js');
+const roomId = 10;
+const botName = 'test';
+
+global.user = '{&quot;email&quot;:&quot;test@test.com&quot;}';
+global.window = { document: { }, location: { href: '' } };
+
+describe('getOrInitializeBotData function >', () => {
+  it('Ok, BotData exists, just retrieve it', (done) => {
+    const testStr = 'testing';
```
Comment (cassiodias) [inline tests/getOrInitializeBotData.js]: As a good practice, we can use expectedResponse or expected{...Something descriptive}

File: tests/getOrInitializeBotData.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    });
+    bdk.__get__('module.exports')(config)
+      .getOrInitializeBotData(roomId, botName, 'test', {})
+      .then((res) => {
+        expect(JSON.parse(res)).to.equal(testStr);
+        done();
+      });
+  });
+
+  it('Ok, BotData does not exist, create it', (done) => {
```
Comment (cassiodias) [inline tests/getOrInitializeBotData.js]: Might be better explaining better 'create it' using default definition or something like that - when I was reading the test was not that clear what was the meaning of create it

File: tests/getOrInitializeBotData.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      return new Promise((resolve) => {
+        resolve({ body:
+          { 'name': 'test2',
+            'value': {}
+          }
+        });
+      });
+    });
+    bdk.__get__('module.exports')(config)
+      .getOrInitializeBotData(roomId, botName, 'test2', '')
```
Comment (cassiodias) [inline tests/getOrInitializeBotData.js]: Here is an important piece of the test and it's just an empty string. Maybe creating a variable describing the default definition to be created and added as well on the expected line (ln 58).

File: tests/getOrInitializeBotData.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      .then((res) => {
+        expect(res)
+          .to.equal('');
+        done();
+      });
+  });
+});
```
Comment (cassiodias) [inline tests/getOrInitializeBotData.js]: Is there any negative case to test here? We have just happy path here... what happens refocus-bdk-client.js line 730 if there is no body?

**Comment by cassiodias** [comment]
Would be helpful having a description e/or a better PR name.

File: tests/getOrInitializeBotData.js (hunk 6 above)
Comment (BrandonMurnane) [reply tests/getOrInitializeBotData.js]: this is most likely to be the default definition

Comment (cassiodias) [reply tests/getOrInitializeBotData.js]: Yes, that is under test and would be better giving the last parameter a better spotlight. 
Just looking the test was not clear to me what is the value of it just looking an empty argument at the end of the method.

File: refocus-bdk-client.js (hunk 3 above)
Comment (kriscfoster) [reply refocus-bdk-client.js]: Maybe we should also be returning the value from the response from refocus, not just what the user gave as an argument

Comment (BrandonMurnane) [reply refocus-bdk-client.js]: This is the same pattern all of the bots have been doing for their data when its initialised.

**Review by kriscfoster — APPROVED** [review]
@BrandonMurnane conflicts

**Review by cassiodias — APPROVED** [review]
There are some questions not addressed (like Tausif 27 days ago) - can you close the conversation giving a reply? Thanks

**Pull request merged**